### PR TITLE
Transform XBRL core_ferc714__yearly_planning_area_demand_forecast table

### LINKED ISSUE #3837: Write transform function to clean and normalize FERC 714 XBRL annual 10-year load forecast table

### PR DESCRIPTION
**Pull request by aesharpe**
# Overview

Closes #3837 .

## What problem does this address?

Adds XBRL data (2021 +) for the annual demand forecast table.

## What did you change?

- Add a `YearlyPlanningAreaDemandForecast` class and class functions.
- Migrate CSV transforms into these class functions.
- Add `forecast` to the column names with forecasted values.

## Ideas
We could add some tests for the 714 data that made sure the mwh and mw values fell within a reasonable range. Might be something we put on the backburner for now.

# Testing

## How did you make sure this worked? How can a reviewer verify this?

Materialize the `core_ferc714__yearly_planning_area_demand_forecast` table in dagster.

```[tasklist]
# To-do list
- [ ] If updating analyses or data processing functions: make sure to update or write data validation tests (e.g. `test_minmax_rows()`)
- [ ] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/nightly/release_notes.html): reference the PR and related issues.
- [ ] Run `make pytest-coverage` locally to ensure that the merge queue will accept your PR.
- [ ] Review the PR yourself and call out any questions or issues you have
- [ ] For minor ETL changes or data additions, once `make pytest-coverage` passes, make sure you have a fresh full PUDL DB downloaded locally, materialize new/changed assets and all their downstream assets and [run relevant data validation tests](https://catalystcoop-pudl.readthedocs.io/en/nightly/dev/testing.html#data-validation) using `pytest` and `--live-dbs`.
- [ ] For bigger ETL or data changes run the full ETL locally and then [run the data validations](https://catalystcoop-pudl.readthedocs.io/en/nightly/dev/testing.html#data-validation) using `make pytest-validate`.
- [ ] Alternatively, run the `build-deploy-pudl` GitHub Action manually.
```

### CODE REVIEW

File: src/pudl/transform/ferc714.py [hunk 1]
```diff
@@ -771,70 +781,152 @@ def out_ferc714__hourly_planning_area_demand(
...
-@asset(
-    io_manager_key="pudl_io_manager",
-    compute_kind="pandas",
-)
-def core_ferc714__yearly_planning_area_demand_forecast(
-    raw_ferc714_csv__yearly_planning_area_demand_forecast: pd.DataFrame,
-) -> pd.DataFrame:
-    """Transform the yearly planning area forecast data per Planning Area.
+class YearlyPlanningAreaDemandForecast:
+    """Class for building the :ref:`core_ferc714__yearly_planning_area_demand_forecast` asset.
```
Comment (aesharpe) [inline src/pudl/transform/ferc714.py:785]: The hourly table has an `out` prefix. Which one should we have here?

**Review by aesharpe — COMMENTED** [review]
Just a high-level comment, but I spent some time tackling the duplicate PK values in the XBRL data before I realized that they were getting handled in the `FercXBRLSQLiteIOManager` class `load_input` function. Specifically, I didn't realize that the older `publication_date` values were getting dropped and the `report_year` column was getting made there. It's a little confusing to have that happen in the IOManager vs. the transforms. I understand that it might be more efficient to do it that way though. 

Out of scope to change for this PR, just a thought I had.

File: src/pudl/transform/ferc714.py (hunk 1 above)
Comment (cmgosnell) [reply src/pudl/transform/ferc714.py:785]: the hourly table is weird bc its so big we will probably never make a de-normalized version of it. I personally think it makes sense to change it to `core_`. But there is a sounds argument for calling it `out_` bc its a table we want users to interact with. ([see comment](https://github.com/catalyst-cooperative/pudl/issues/3809#issuecomment-2342019024))

which is all to say i think this one should stay as `core_`

File: src/pudl/metadata/fields.py [hunk 2]
```diff
@@ -2680,9 +2680,9 @@
         ),
         "unit": "MW",
     },
-    "net_demand_mwh": {
+    "net_demand_forecast_mwh": {
         "type": "number",
-        "description": "Net electricity demand for the specified period in megawatt-hours (MWh).",
+        "description": "Net forecasted electricity demand for the specific period in megawatt-hours (MWh).",
```
Comment (cmgosnell) [inline src/pudl/metadata/fields.py:2685]: I love the addition of forecast in here. it makes it much more clear.

not blocking... but is there a difference between this and `demand_forecast_mwh`?

File: src/pudl/transform/ferc714.py [hunk 3]
```diff
@@ -771,70 +781,161 @@ def out_ferc714__hourly_planning_area_demand(
...
+        This function fixes the following errors:
+
+        - There's one record with an NA forecast_year value. This row
+          also has no demand forcast values. Because forcast_year is a primary key
+          we can't have any NA values. Because there are no substantive forcasts
+          in this row, we can safely remove this row.
+        - respondent_id_ferc714 number 107 reported their forecast_year
+          as YY instead of YYYY values.
+        - There's also at least one forecast year value reported as 3022 that should
+          be 2033.
```
Comment (cmgosnell) [inline src/pudl/transform/ferc714.py:846]: lol

File: src/pudl/transform/ferc714.py [hunk 4]
```diff
@@ -771,70 +781,161 @@ def out_ferc714__hourly_planning_area_demand(
...
+        mask = (
+            (df["respondent_id_ferc714"] == 17)
+            & (df["report_year"] == 2023)
+            & (df["forecast_year"] == 3033)
+        )
+        df.loc[mask, "forecast_year"] = 2033
+        # Make sure forecast_year values are expected
+        assert (
+            df["forecast_year"].isin(range(2021, 3001)).all()
+        ), "Forecast year values not in expected range"
```
Comment (cmgosnell) [inline src/pudl/transform/ferc714.py:872]: the forecasts go through 3000 eeep really?!

File: src/pudl/transform/ferc714.py [hunk 5]
```diff
@@ -771,70 +781,161 @@ def out_ferc714__hourly_planning_area_demand(
...
+                    "net_demand_forecast_mwh",
+                ]
+            ]
+            .mean()
+            .reset_index()
+        )
+        # Make sure no more rows were dropped than expected
+        assert (
+            original_len - len(df) == 21
+        ), f"dropped {original_len - len(df)} rows, expected 26"
```
Comment (cmgosnell) [inline src/pudl/transform/ferc714.py:914]: nit but these expected numbers are off

File: src/pudl/transform/ferc714.py [hunk 6]
```diff
@@ -771,70 +781,161 @@ def out_ferc714__hourly_planning_area_demand(
...
+            df.groupby(["respondent_id_ferc714", "report_year", "forecast_year"])[
+                [
+                    "summer_peak_demand_forecast_mw",
+                    "winter_peak_demand_forecast_mw",
+                    "net_demand_forecast_mwh",
+                ]
+            ]
+            .mean()
+            .reset_index()
+        )
```
Comment (cmgosnell) [inline src/pudl/transform/ferc714.py:910]: completely non-blocking nit/i know you didn't introduce this. I personally like doing a split/apply/combine process when altering only a few records. in theory this `mean` will do nothing to a group of 1. but i like to stay cautious about that.

```python
dupe_mask = {soemthing}
deduped = df[dupe_mask].groupby()...stuff
pd.concat([df[~dupe_mask], deduped]) 
```

**Review by cmgosnell — CHANGES_REQUESTED** [review]
this looks grand!! i have two real but minor questions and one nittiest of nit.

File: src/pudl/metadata/fields.py (hunk 2 above)
Comment (aesharpe) [reply src/pudl/metadata/fields.py:2685]: Yes, `demand_forecast_mwh` is from 930 and is a day-ahead forecast. This is a year-ahead forecast. The 930 value is also demand vs. the 714 value which is net demand.

Comment (aesharpe) [reply src/pudl/metadata/fields.py:2685]: We could probably add a day-ahead suffix or something to the 930 value! What do you think?

Comment (aesharpe) [reply src/pudl/metadata/fields.py:2685]: _Technically_ for the FERC714 table it's net_load vs net_demand, but they are sooo similar idk if it matters. But it's called net load in the FERC taxonomy.

File: src/pudl/transform/ferc714.py (hunk 4 above)
Comment (aesharpe) [reply src/pudl/transform/ferc714.py:872]: No haha I was just trying to think of a reasonable range. We could certainly lower this! I was just not sure if it was something we wanted to update every year or use as a check for obvious outliers.

File: src/pudl/transform/ferc714.py (hunk 6 above)
Comment (aesharpe) [reply src/pudl/transform/ferc714.py:910]: Makes sense! I sometimes fear that I'll introduce more problems with split/apply/combine than just applying something to an intact database. But I should trust my code more!

File: src/pudl/transform/ferc714.py (hunk 4 above)
Comment (aesharpe) [reply src/pudl/transform/ferc714.py:872]: I think I can probably safely lower it to 2100

File: src/pudl/transform/ferc714.py (hunk 5 above)
Comment (aesharpe) [reply src/pudl/transform/ferc714.py:914]: idk what that was about haha but fixed!

File: src/pudl/metadata/fields.py (hunk 2 above)
Comment (cmgosnell) [reply src/pudl/metadata/fields.py:2685]: yea it makes sense to me to stay true to the source taxonomy unless we learn its wrong somehow, but that's not the case here

**Review by cmgosnell — APPROVED** [review]
🎉

**Comment by cmgosnell** [comment]
oh! @aesharpe ! can you add the new years of data into the `Ferc714CheckSpec` asset check?

**Pull request merged**